core/chains/nano.go

```diff
@@ -0,0 +1,177 @@
+package chains
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"math/big"
+	"net/http"
+	"sort"
+)
+
+func nanoRequest[ReqType any, ResType any](url string, req ReqType) (*ResType, error) {
+	var response ResType
+	body, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := http.Post(url, "application/json", bytes.NewBuffer(body))
+	if err != nil {
+		return nil, err
+	}
+	body, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if err = json.Unmarshal(body, &response); err != nil {
+		return nil, err
+	}
+	return &response, nil
+}
+
+func Nano() (int, error) {
+	rpc := "http://localhost:7076"
+	richlist := "https://nano.nendly.com/nakamoto/richlist"
+
+	bigFromStr := func(value string) *big.Int {
+		bigValue, _ := new(big.Int).SetString(value, 10)
+		return bigValue
+	}
+
+	// Determine the amount of nano that's needed for an attack
+	type OnlineRequest struct {
+		Action string `json:"action"`
+	}
+	type OnlineResponse struct {
+		OnlineStake string `json:"trended_stake_total,omitempty"`
+		Error       string `json:"error,omitempty"` // nano nodes return status 200 even when erroring
+	}
+	onlineRes, err := nanoRequest[OnlineRequest, OnlineResponse](rpc, OnlineRequest{
+		Action: "confirmation_quorum",
+	})
+	if onlineRes != nil && onlineRes.Error != "" {
+		log.Println(onlineRes.Error)
+		return -1, errors.New("failed to get the online stake")
+	}
+	if err != nil {
+		return -1, err
+	}
+	online := bigFromStr(onlineRes.OnlineStake)
+	needed := new(big.Int).Div(online, big.NewInt(3))
+
+	// Request the list of reps
+	type RepsRequest struct {
+		Action  string `json:"action"`
+		Sorting bool   `json:"sorting"`
+		Count   int    `json:"count"`
+	}
+	type RepsResponse struct {
+		Reps  map[string]string `json:"representatives,omitempty"`
+		Error string            `json:"error,omitempty"`
+	}
+	repsRes, err := nanoRequest[RepsRequest, RepsResponse](rpc, RepsRequest{
+		Action:  "representatives",
+		Sorting: true,
+		Count:   100,
+	})
+	if repsRes != nil && repsRes.Error != "" {
+		log.Println(repsRes.Error)
+		return -1, errors.New("failed to get the list of representatives")
+	}
+	if err != nil {
+		return -1, err
+	}
+	reps := make(map[string]*big.Int)
```

Comment:
what does `reps` represent? repetitions/iterations ?..

Reply:
maybe give it a more qualified name indicating what it is intended to do

Reply:
In nano, validators are called `representatives`. I went ahead and changed `reps` to `validators` to avoid confusion.

Reply:
Cool. Maybe add a comment about this?

Reply:
I've added a comment explaining it for clarity